src/lib.rs

```diff
@@ -461,143 +720,388 @@ mod tests {
-    );
+    let res = recursive_snark.verify(&pp, 3, &v1, &v2);
     assert!(res.is_ok());
+
+    // verification fails when num_steps is incorrect
+    let bad_res = recursive_snark.verify(&pp, 2, &v1, &v2);
+    assert!(bad_res.is_err());
   }
 
   #[test]
   fn test_ivc_nontrivial() {
     // produce public parameters
-    let pp = PublicParams::<
-      G1,
-      G2,
-      TrivialTestCircuit<<G1 as Group>::Scalar>,
-      CubicCircuit<<G2 as Group>::Scalar>,
-    >::setup(
-      TrivialTestCircuit {
-        _p: Default::default(),
-      },
-      CubicCircuit {
-        _p: Default::default(),
-      },
+    let pp = PublicParams::<G1, G2, TrivialTestCircuit<S1>, CubicCircuit<S2>, U1, U1>::setup(
+      TrivialTestCircuit::default(),
+      CubicCircuit::default(),
     );
 
     let num_steps = 3;
 
     // produce a recursive SNARK
     let res = RecursiveSNARK::prove(
       &pp,
-      num_steps,
-      <G1 as Group>::Scalar::one(),
-      <G2 as Group>::Scalar::zero(),
+      Some(num_steps),
+      &IO::Val(S1::one()),
+      &IO::Val(S2::zero()),
+      None,
+      None,
     );
     assert!(res.is_ok());
     let recursive_snark = res.unwrap();
 
+    let secondary_z0 = S2::zero();
+    let v1 = pp.new_primary_val(S1::one());
+    let v1_bad = pp.new_primary_val(S1::zero());
+    let v2 = pp.new_secondary_val(secondary_z0);
+
     // verify the recursive SNARK
-    let res = recursive_snark.verify(
-      &pp,
-      num_steps,
-      <G1 as Group>::Scalar::one(),
-      <G2 as Group>::Scalar::zero(),
-    );
+    let res = recursive_snark.verify(&pp, num_steps, &v1, &v2);
     assert!(res.is_ok());
 
+    // verification fails when num_steps is incorrect
+    let bad_res = recursive_snark.verify(&pp, num_steps + 1, &v1_bad, &v2);
+    assert!(bad_res.is_err());
+
     let (zn_primary, zn_secondary) = res.unwrap();
 
     // sanity: check the claimed output with a direct computation of the same
-    assert_eq!(zn_primary, <G1 as Group>::Scalar::one());
-    let mut zn_secondary_direct = <G2 as Group>::Scalar::zero();
+    assert_eq!(zn_primary, S1::one());
+    let mut zn_secondary_direct = secondary_z0;
     for _i in 0..num_steps {
       zn_secondary_direct = CubicCircuit {
         _p: Default::default(),
       }
-      .compute(&zn_secondary_direct);
+      .compute(&zn_secondary_direct)
+      .unwrap()
+      .1;
     }
     assert_eq!(zn_secondary, zn_secondary_direct);
-    assert_eq!(zn_secondary, <G2 as Group>::Scalar::from(2460515u64));
+    assert_eq!(zn_secondary, S2::from(2460515u64));
   }
 
   #[test]
-  fn test_ivc_base() {
+  fn test_ivc_binary() {
     // produce public parameters
-    let pp = PublicParams::<
-      G1,
-      G2,
-      TrivialTestCircuit<<G1 as Group>::Scalar>,
-      CubicCircuit<<G2 as Group>::Scalar>,
-    >::setup(
-      TrivialTestCircuit {
-        _p: Default::default(),
-      },
-      CubicCircuit {
+    let p = PoseidonConstants::<S2, U2>::new();
```

Comment:
Here is an example of the caller/user needing to explicitly allocate PoseidonConstants.